Editor: Remove sticky option if post is private or password protected

jeremeylduvall opened the pull request:
This is an attempt to fix #13446. Right now, the sticky option appears on posts that are private or password protected. Per the original issue, this option is not available in wp-admin or in core.

I've labeled this as "In Progress" as I would like some feedback particularly on the tests added (described below).

A few notes:
- We set `sticky` to `false` when a post visibility is updated to either private or password protected. This means that currently "sticky" posts that are private or password protected will continue to be sticky until the visibility is flipped. Per the original report, it seems like there's some operation that removes the sticky flag from these posts over time as well.
- It was necessary to add the `export` statement [here](https://github.com/Automattic/wp-calypso/blob/2513d535bccfae77b46ebb92130b6031b44c878e/client/post-editor/edit-post-status/index.jsx#L30) for the tests to be run without the Redux store. Per the [Redux docs](http://redux.js.org/docs/recipes/WritingTests.html), there shouldn't be any adverse side effects since the default export will work elsewhere.
- I wrote some minor tests to make sure we have the correct behavior. With these tests though, I'm running into two errors that I'm not quite sure how to solve and would love some feedback:

> Warning: Failed prop type: Invalid prop `require` of type `string` supplied to `AsyncLoad`, expected `function`.`

I assume this is happening due to the `AsyncLoad` for [`components/post-schedule`](https://github.com/Automattic/wp-calypso/blob/2513d535bccfae77b46ebb92130b6031b44c878e/client/post-editor/edit-post-status/index.jsx#L234). I'm flummoxed on the workaround though.

> double callback!

The tests emit a double callback, but for the life of me, I can't figure out why/where. Would love another set of eyes. See relevant previous discussion in #4206 about this behavior.

## To Test
1. Load this branch and open a new post at http://calypso.localhost:3000/post/.
2. Under "Status" in the sidebar, try setting the post to password protected. The sticky option should disappear.
3. Set it back to public. The sticky option should reappear.
4. Publish it as a private post. The sticky option will disappear.

To test for regressions, visit wordpress.com/post and publish a private post as "sticky." Then, open that post in the branch. You should not see any regressions.

To run the tests, use: `npm run test-client client/post-editor/edit-post-status/test`

## GIF
![sticky](https://cloud.githubusercontent.com/assets/7240478/25689382/c7b1ca72-3044-11e7-8e9b-ba4c5214c90e.gif)

bperson commented:
We should also probably disable sticking the post via the editor's action bar when it's password-protected/private?

Vertical spacing seems a bit off when the option is hidden:
<img width="224" alt="screen shot 2017-05-04 at 08 51 01" src="https://cloud.githubusercontent.com/assets/1145270/25692959/8dc077f2-30a6-11e7-99bb-138cc9ffa8c3.png">

Last nitpick: the "private" visibility: why does it require to update the post to do anything? It seems password-protected is better handled? (this is most likely unrelated to this PR though ;) )

youknowriad commented:
> Warning: Failed prop type: Invalid prop require of type string supplied to AsyncLoad, expected function.`

I think this is happening because the `AsyncLoad` component needs the babel-plugin-transform-wpcalypso-async and Webpack code splitting to work properly and I assume we're not using those for unit tests. @aduth any hints here?

youknowriad commented:
This is working as expected but as noted by @bperson It seems that the "sticky" toggle on the sidebar relies on the "saved" post stats to hide/show the toggle. Maybe it should rely on the "edited" post instead.

jeremeylduvall commented:
> We should also probably disable sticking the post via the editor's action bar when it's password-protected/private?

Good call. I actually didn't know that was possible tbh. I thought the flag just appeared when you used the toggle. I didn't realize it was a toggle as well. Will address. Along with:

> Vertical spacing seems a bit off when the option is hidden:

For this point:

> the "private" visibility: why does it require to update the post to do anything?

That's a good point, but probably a discussion that would need to start with core first. "Private" is basically "Privately published." In wp-admin, it looks like the "Save Draft" button disappears when I select private so I have to update. I'm not sure of the reason.

aduth commented:
>I think this is happening because the `AsyncLoad` component needs the babel-plugin-transform-wpcalypso-async and Webpack code splitting to work properly and I assume we're not using those for unit tests. @aduth any hints here?

We don't have a good solution here yet aside from avoiding having a file under test require one which depends on a Webpack loader.

See: #12203

aduth commented on client/post-editor/edit-post-status/test/index.jsx:
```diff
@@ -0,0 +1,46 @@
+	useFakeDom();
+
+	before( function() {
+		shallow = require( 'enzyme' ).shallow;
```
Enzyme doesn't depend on the fake DOM, so could be moved to top-level imports.

jeremeylduvall commented:
Updated this to address the points above. A few notes:
- I had to connect the `EditorActionBar` component (and as a result convert to ES6) to actively respond to the post state and show/hide `EditorSticky`. Otherwise, the `EditorSticky` would only react to changes when the post was actively saved/updated. This way, it reacts whenever you change the post visibility.
- I swapped around the styling so that everything looks good with the sticky option removed. Here's a visual comparison when sticky is present vs. removed.

<img width="271" alt="screen shot 2017-05-05 at 2 26 10 pm" src="https://cloud.githubusercontent.com/assets/7240478/25763127/d67935da-319e-11e7-98c2-03feb6b3220b.png">

<img width="273" alt="screen shot 2017-05-05 at 2 16 12 pm" src="https://cloud.githubusercontent.com/assets/7240478/25763132/da640364-319e-11e7-8b6b-c7d6e0618ae2.png">

@youknowriad to this point:

> It seems that the "sticky" toggle on the sidebar relies on the "saved" post stats to hide/show the toggle. Maybe it should rely on the "edited" post instead.

I'm not sure I understand. Both sticky elements (the toggle and bookmark component in the action bar) should be removed as soon as you change the post visibility (no save required). Could you clarify?

youknowriad commented:
> I'm not sure I understand. Both sticky elements (the toggle and bookmark component in the action bar) should be removed as soon as you change the post visibility (no save required). Could you clarify

Actually, I was saying the same thing you said about the `EditorActionBar` change. The `Sticky` component was not responding to the edits directly but was changing only when post was saved/updated. It seems fixed now.

---

I'm still seeing some undesired bahaviour while the `sticky` component is not shown on status change. Here are the steps to reproduce:

 * Publish privately a post
 * Switch the post status to "public"
 * The `Sticky` component is not shown until I effectlively update/save the post.

youknowriad commented on client/post-editor/editor-action-bar/index.jsx:
```diff
@@ -14,28 +15,29 @@ import Tooltip from 'components/tooltip';
 	render() {
 		const multiUserSite = this.props.site && ! this.props.site.single_user_site;
+		const isPasswordProtected = utils.getVisibility( this.props.post ) === 'password';
+		const isPrivate = utils.isPrivate( this.props.post );
```
It looks like you're relying on the Redux "Edited Post" to show/hide the `EditorSticky` component. But it looks like, we're still relying on the Flux store for "status" updates (see https://github.com/Automattic/wp-calypso/blob/fix/13446-private-posts-sticky/client/post-editor/editor-visibility/index.jsx#L181). This makes me think we may want to rely on the Flux "Edited Post" instead?

youknowriad replied:
I've tried and failed to reduxify the `status` updates in #12353 due to the multiple possible side effects.

jeremeylduvall commented:
Good catch. We do indeed need to use the Flux post to monitor status changes. Updated now. The behavior you described seems to have been fixed.

I also added 8px of padding to `.edit-post-status__revert-to-draft`. This way, the visibility doesn't cramp the button.

Original:
<img width="283" alt="screen shot 2017-05-08 at 7 27 52 am" src="https://cloud.githubusercontent.com/assets/7240478/25806561/da27ed9c-33c0-11e7-9dd9-3b2f65b2b61d.png">

Updated:
<img width="309" alt="screen shot 2017-05-08 at 7 27 36 am" src="https://cloud.githubusercontent.com/assets/7240478/25806554/d154ec42-33c0-11e7-8b6f-a9734c1716a0.png">

What are your thoughts about the distance between the date and "Revert to Draft" when the sticky toggle is missing? I'm mixed on whether or not it's too large. We could adjust the top margin on the revert to draft button.

youknowriad commented on client/post-editor/editor-action-bar/index.jsx:
```diff
@@ -14,28 +15,30 @@ import Tooltip from 'components/tooltip';
 
 	render() {
 		const multiUserSite = this.props.site && ! this.props.site.single_user_site;
+		const isPasswordProtected = utils.getVisibility( this.props.post ) === 'password';
```
Should we update the `isPasswordProtected` to use the Flux post as well? I don't know if it's possible but it feels a bit weird to check flux for the "isPrivate" and Redux for the "isPasswordProtected". These migrations are hard :(

jeremeylduvall replied:
@youknowriad It does indeed feel weird :( It looks like we're storing password changes in Redux though ([example](https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/editor-visibility/index.jsx#L180)) so I think we need to stick with `this.props.post`. Changing to the Flux post causes some unexpected behavior when you change from password -> public -> password.

youknowriad replied:
Let's leave this as is, but maybe we should add a comment to explain why we do this?

youknowriad commented on client/post-editor/editor-action-bar/index.jsx:
```diff
@@ -82,4 +88,18 @@ export default React.createClass( {
+			post
+		};
+	},
+)( EditorActionBar );
```
I'm seeing another regression due to the reduxification here. The "Author" switcher is not updating the author correctly when we select a different one until we save. (That's probably because it's receiving the Redux Author while it performs the changes on Flux)

jeremeylduvall commented:
As we discussed in Slack, the author is pretty messy due to the Flux vs. Redux data storage. This should be now fixed by monitoring the `PostEditStore` and reacting to the change event. For reasons I'm not quite sure of, `PostEditStore.get()` returns `null` during the initial render, which is why we're using `props.post` as a fallback value.

I also made some tweaks so the status area should stay consistent when the toggle is missing. Here's a GIF of the various flows showing the style changes:

![status](https://cloud.githubusercontent.com/assets/7240478/25949898/6303de08-3616-11e7-9b2a-1a299beb2c70.gif)

Related to our conversation, I'm torn on whether we need to repeat the same listener pattern in the other areas where we're relying on the Flux store for data (dealing with visibility updates). Right now, it seems like changes at the `PostEditor` component are re-rendering the status and action bar correctly.

youknowriad commented:
> Right now, it seems like changes at the PostEditor component are re-rendering the status and action bar correctly.

Right, this happens because the `PostEditor` is a parent component but it makes the sub-components that rely on `PostEditStore` not reusable. I'd prefer to watch for the changes in each one of those. In the same time, this adds more Flux stores in our components and this annoys me a bit. @aduth Any ideas on the issues here.

youknowriad commented:
> If I understand correctly, they're reusable in the sense that as long as the parent component passes in the correct props, it will work correctly.

Problem is, the parent is not passing the prop, they're automatically accessing `PostEditStore.get()` without watching the store.

youknowriad commented:
Oh sorry for the confusion, there's two components on the current PR `EditPostStatus` and `EditorActionBar`

aduth commented:
Oh, I see now. Yes, the general guideline would be that if the component calls to `PostEditStore.get()` in its render logic, it **must** bind to the store to ensure that updates are reflected. Pretty miserable that we need to source from both Flux and Redux post though 😞  We could pass in `visibility` / `isPrivate` from an ancestor component which has those bindings already (e.g. `PostEditor`), but this goes to your point of consuming at the point they're needed. There's not a great solution here until more of data is managed exclusively through Redux. I'd probably lean more on avoiding `PostEditStore.on` in favor of passing in the Flux post (or derived properties) from ancestor components.

jeremeylduvall commented:
Got it. The two issues then are in [`EditorActionBar`](https://github.com/Automattic/wp-calypso/blob/469de4be87785b43521016ba9d5dc54d3c625731/client/post-editor/editor-action-bar/index.jsx#L44) and [`EditPostStatus`](https://github.com/Automattic/wp-calypso/blob/469de4be87785b43521016ba9d5dc54d3c625731/client/post-editor/edit-post-status/index.jsx#L109) where we're accessing `PostEditStore.get()` in both cases to determine whether the post is private without monitoring the store.

I'll take a look at how many props we would need to pass down (and how many levels) if we just relied on `PostEditor` and the existing Flux bindings there to feed the various data we need from Flux.

jeremeylduvall commented:
I  updated this one today. All necessary Flux data should be passed down from the parent (`PostEditor`) so we're no longer relying on `PostEditStore.get()`. The post privacy (labeled as `isPostPrivate` to distinguish from sections where we use `isPrivate` to refer to the site) and the post author are the two bits of info needed.

There shouldn't be any regressions in behavior. Author switcher and post privacy changes should all work correctly in the editor.
